chore: Add Dockerfile and publish workflow

poppoerika opened the pull request:
momento-proxy Docker image is tested on an EC2 instance and I was able to set a key/value pair.
```
[root@ip-172-31-53-115 bin]# telnet 0.0.0.0 11211
Trying 0.0.0.0...
Connected to 0.0.0.0.
Escape character is '^]'.
set erika 0 0 3
eri 
STORED
```

Every time a pr is merged to `main`, a new Docker image is built and pushed to the `gomomento/momento-proxy` repo.
When users run `docker run`, they need to provide an env variable `MOMENTO_AUTHENTICATION` (Momento auth token).
Also updated `momento-proxy.md` to include how to run the momento-proxy Docker image.

eaddingtonwhite commented on src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:
```diff
@@ -198,7 +198,7 @@ fn parse_get(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
                         #[cfg(feature = "stats")]
                         GET_CARDINALITY.increment(PreciseInstant::now(), keys.len() as _, 1);
-                        
+
```
Id just revert this file so your commit isnt modifying any rust files just so easier to review when we contribute down.

poppoerika replied:
@eaddingtonwhite https://github.com/brayniac/pelikan/blob/momento-proxy/src/rust/protocol/memcache/src/memcache/wire/request/parse.rs#L201
The original file has a new line on 201. So added it back.

eaddingtonwhite commented on src/rust/proxy/momento/README.md:
```diff
@@ -41,18 +41,34 @@ use with the proxy.
 The resulting binary will be `target/release/momento_proxy` and it can be copied
 to a standard system path (eg: `/usr/local/bin`).
+
+### momento-proxy Docker image
```
we should have instructions here for how to build, run, and test container locally as well

eaddingtonwhite commented on src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:
```diff
@@ -198,7 +198,6 @@ fn parse_get(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
                     } else {
                         #[cfg(feature = "stats")]
                         GET_CARDINALITY.increment(PreciseInstant::now(), keys.len() as _, 1);
-                        
```
I would revert this line

poppoerika replied:
Ah... I think VSCode formatted it on save.

eaddingtonwhite commented on src/rust/proxy/momento/README.md:
````diff
@@ -41,18 +41,66 @@ use with the proxy.
+- Running the newly built image with the default config:
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN>
````
are we missing image name in this? Doesnt `momento-proxy` need to be in command somewhere?

eaddingtonwhite commented on src/rust/proxy/momento/README.md:
````diff
@@ -41,18 +41,66 @@ use with the proxy.
+
+```
+docker pull gomomento/momento-proxy
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> momento-proxy
````
does this need to be `gomomento/momento-proxy` for image to run? vs  just `momento-proxy` below for local y built version?

poppoerika replied:
Yup, I confirmed that it needs to be `gomomento/momento-proxy` to pull the image from our DockerHub repo. 
I think we might be able to do `momento-proxy` with a paid subscription (business) to DockerHub.

eaddingtonwhite replied:
So should it be like this now?
```suggestion
docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> gomomento/momento-proxy
```

eaddingtonwhite commented on src/rust/proxy/momento/README.md:
````diff
@@ -41,18 +41,66 @@ use with the proxy.
+To set your own, please provide an env variable `CONFIG` when running a container.
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> -e CONFIG=<YOUR_CONFIG_FILE> momento-proxy
````
Same here should this be like:
```suggestion
docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> -e CONFIG=<YOUR_CONFIG_FILE> gomomento/momento-proxy
```

eaddingtonwhite commented on src/rust/proxy/momento/README.md:
```diff
@@ -41,18 +41,66 @@ use with the proxy.
+
+### momento-proxy Docker image
+
+Running the momento-proxy image to run the Momento proxy.
```
Maybe reword lik:
```suggestion
You can run the `momento-proxy` container by pulling it from [Momento's](https://momentohq.com/) [docker hub registry](https://hub.docker.com/u/gomomento) with following commands. 
```

poppoerika commented on src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:
```diff
@@ -198,7 +198,7 @@ fn parse_get(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
                         #[cfg(feature = "stats")]
                         GET_CARDINALITY.increment(PreciseInstant::now(), keys.len() as _, 1);
-                        
+
```
I think I added it back? 
I see a new line in my file...
<img width="803" alt="image" src="https://user-images.githubusercontent.com/22686260/165629014-1994adf5-61c1-475c-9c18-3e8822647591.png">

eaddingtonwhite replied:
Its more then new line though its a new line plus some white space 😬. You can see in diff my suggestion block ☝️

poppoerika replied:
💡

eaddingtonwhite reviewed (approved):
@poppoerika This is looking good to me nice work! Lets get it merged and can do final verification again in our main `momento` branch make sure image gets published properly. Then can work to get contribute branch down into Bryan's fork. 

Maybe have @danielamiao take one more pass on content quick just to get one more pair eyes.

danielamiao commented on src/rust/proxy/momento/README.md:
````diff
@@ -41,18 +41,66 @@ use with the proxy.
+```
+
+By default, [this configuration](../../../../config/momento_proxy.toml) is used for the Momento proxy.
+To set your own, please provide an env variable `CONFIG` when running a container.
````
My Dockerfile knowledge is limited, but have you tried this with a custom configuration file on EC2 instance? In the above Dockerfile I see a line where you are setting the default `CONFIG` value, and a line for copying the default config `momento_proxy.toml`, then a run command line with `CONFIG` env. However, if I supply a custom CONFIG path, how does that get copied into the Docker container?

poppoerika replied:
I think `-v` needs to be passed with `docker run` to mount a volume to a Docker container. 
I'll test it out first and update the README accordingly.

poppoerika commented:
@eaddingtonwhite @danielamiao 
Updated the Dockerfile a bit so that we can mount a host machine's config directory to a container's config directory to be able to use a custom config file passed by users.
Also updated README to reflect this change.
Tested it on an EC2 instance and was able to use my custom config file to run.

eaddingtonwhite commented on src/rust/proxy/momento/README.md:
````diff
@@ -41,18 +41,66 @@ use with the proxy.
+- Running the newly built image with your custom config:
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> -e CONFIG=<YOUR_CONFIG_FILE> momento-proxy
````
does this need to be updated to have a `-v` flag passed for mounting config file?

eaddingtonwhite reviewed (approved):
Nice work @poppoerika
